src/ALE/Recon1d_PLM_WLS.F90

```diff
@@ -0,0 +1,460 @@
+         max(abs(2. * this%u_mean(k)), abs(this%ul(k)), abs(this%ur(k))) ) check_reconstruction = .true.
+  enddo
+
+  ! Create a perturbable reconstruction
+  call perturbed%init( this%n, h_neglect=this%h_neglect )
+  call perturbed%reconstruct( h, u ) ! Should reproduce "this"
+  ! Check the copy is identical
+  do k = 1, this%n
+    if ( abs( perturbed%u_mean(k) - this%u_mean(k) ) > 0. ) check_reconstruction = .true.
+    if ( abs( perturbed%ul(k) - this%ul(k) ) > 0. ) check_reconstruction = .true.
+    if ( abs( perturbed%ur(k) - this%ur(k) ) > 0. ) check_reconstruction = .true.
+    if ( abs( perturbed%slp(k) - this%slp(k) ) > 0. ) check_reconstruction = .true.
+  enddo
+  ! Now perturb the slope. The local error should not decrease.
+  do k = 1, this%n
+    slp = this%slp(k) * ( 1.0 + 1. * epsilon(slp) )
+    perturbed%slp(k) = slp
+    perturbed%ul(k) = u(k) - 0.5 * slp
+    perturbed%ur(k) = u(k) + 0.5 * slp
+    if ( LS_error(perturbed, k, h, u) < LS_error(this, k, h, u) ) check_reconstruction = .true.
+
+    slp = this%slp(k) * ( 1.0 - 1. * epsilon(slp) )
+    perturbed%slp(k) = slp
+    perturbed%ul(k) = u(k) - 0.5 * slp
+    perturbed%ur(k) = u(k) + 0.5 * slp
+    if ( LS_error(perturbed, k, h, u) < LS_error(this, k, h, u) ) check_reconstruction = .true.
+  enddo
+
+end function check_reconstruction
+
+!> Returns local least squares error for a particular cell
+!!
+!! Note that this is the error relative to the minimum of the loss function so that at the
+!! true solution this function returns zero. See module documentation.
+real function LS_error(this, k, h, u)
+  type(PLM_WLS), intent(in) :: this !< This reconstruction
+  integer,        intent(in) :: k    !< Cell number
+  real,           intent(in) :: h(*) !< Grid spacing (thickness) [typically H]
+  real,           intent(in) :: u(*) !< Cell mean values [A]
+  ! Local variables
+  real :: u_l, u_r, u_c ! Left, right, and center values [A]
+  real :: h_l, h_c, h_r ! Thickness of left, center and right cells [H]
+  real :: h_l0, h_r0, hc0 ! Thickness of left, right, center cells with h_neglect added [H]
+  real :: hx2l, hx2r ! Contributions to denominator, <h x^2> [H3]
+  real :: hxyl, hxyr ! Contributions to numerator, <h x y> [H2 A]
+  integer :: km1, kp1
+
+  km1 = max(1, k-1)
+  kp1 = min(this%n, k+1)
+  u_l = u(km1)
+  u_c = u(k)
+  u_r = u(kp1)
+
+  h_l = h(km1) * real( k - km1 ) ! This zeroes h_l at k==1
+  h_r = h(kp1) * real( kp1 - k ) ! This zeroes h_r at k==n
+  h_c = h(k)
+  hc0 = h_c + this%h_neglect
+
+  h_l0 = h_l + this%h_neglect
+  h_r0 = h_r + this%h_neglect
+  hxyl = ( h_l * 0.5 * ( h_c + h_l ) ) * ( u_c - u_l )
+  hxyr = ( h_r * 0.5 * ( h_c + h_r ) ) * ( u_r - u_c )
+  hx2l = h_l0 * 0.25 * ( h_c + h_l0 )**2
+  hx2r = h_r0 * 0.25 * ( h_c + h_r0 )**2
+  LS_error = h_c * ( ( hx2l + hx2r ) * this%slp(k) - h(k) * ( hxyl + hxyr ) )**2
+  LS_error = LS_error / ( hc0 * ( hx2l + hx2r ) )
```

Comment:
When `hc0`, `hx2l` and `hx2r` are set to `tiny(h_c)` and `tiny(h_c)**2`, which is the effective default behavior when `h` goes to 0 and no value is provided for `h_neglect` in the `this%init()` call, I believe that their product in the denominator will underflow to 0, leading to a `NaN` being returned by `LS_error()`.  I think that `LS_error()` may need better error handling to work when `h` goes to 0.